pkg/storage/migrator/sync.go

```diff
@@ -0,0 +1,238 @@
+		if s.config.Locker_handler != nil {
+			s.config.Locker_handler()
+		} else {
+			s.config.Lockable.Lock()
+		}
+	}
+	conf.Unlocker_handler = func() {
+		if s.config.Unlocker_handler != nil {
+			s.config.Unlocker_handler()
+		} else {
+			s.config.Lockable.Unlock()
+		}
+	}
+	conf.Concurrency = map[int]int{
+		storage.BlockTypeAny: 16,
+	}
+	if s.config.Concurrency != nil {
+		conf.Concurrency = s.config.Concurrency
+	}
+
+	conf.Integrity = s.config.Integrity
+	conf.Cancel_writes = s.config.Cancel_writes
+	conf.Dedupe_writes = s.config.Dedupe_writes
+
+	conf.Progress_handler = func(p *MigrationProgress) {
+		log.Info().
+			Str("name", s.config.Name).
+			Float64("migrated_blocks_perc", p.Migrated_blocks_perc).
+			Int("ready_blocks", p.Ready_blocks).
+			Int("total_blocks", p.Total_blocks).
+			Float64("ready_blocks_perc", p.Ready_blocks_perc).
+			Int("active_blocks", p.Active_blocks).
+			Int("total_migrated_blocks", p.Total_Migrated_blocks).
+			Int("total_canceled_blocks", p.Total_Canceled_blocks).
+			Int("total_duplicated_blocks", p.Total_Duplicated_blocks).
+			Msg("Continuous sync progress")
+		if s.config.Progress_handler != nil {
+			s.config.Progress_handler(p)
+		}
+	}
+	conf.Error_handler = func(b *storage.BlockInfo, err error) {
+		log.Error().
+			Str("name", s.config.Name).
+			Err(err).
+			Int("block", b.Block).
+			Int("type", b.Type).
+			Msg("Continuous sync error")
+		if s.config.Error_handler != nil {
+			s.config.Error_handler(b, err)
+		}
+	}
+
+	// When a block is written, update block_status with the largest ID for that block
+	conf.Block_handler = func(b *storage.BlockInfo, id uint64, data []byte) {
+		hash := sha256.Sum256(data)
+
+		s.block_status_lock.Lock()
+		defer s.block_status_lock.Unlock()
+		if id > s.block_status[b.Block].Current_ID {
+			s.block_status[b.Block].Current_ID = id
+			s.block_status[b.Block].Current_hash = hash
+		}
+	}
+
+	mig, err := NewMigrator(s.config.Tracker, s.config.Destination, s.config.Orderer, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	num_blocks := (s.config.Tracker.Size() + uint64(s.config.Block_size) - 1) / uint64(s.config.Block_size)
+
+	if sync_all_first {
+		// Now do the initial migration...
+		err = mig.Migrate(int(num_blocks))
+		if err != nil {
+			return nil, err
+		}
+
+		// Wait for completion.
+		err = mig.WaitForCompletion()
+		if err != nil {
+			return nil, err
+		}
+
+		if s.config.Hashes_handler != nil {
+			hashes := mig.GetHashes()
+			s.config.Hashes_handler(hashes)
+		}
+	} else {
+		// We don't need to do an initial migration.
+		for b := 0; b < int(num_blocks); b++ {
+			mig.SetMigratedBlock(b)
+		}
+		// Track changes for everything.
+		s.config.Tracker.TrackAt(0, int64(s.config.Tracker.Size()))
+	}
+
+	// Now enter a loop looking for more dirty blocks to migrate...
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			// Context has been cancelled. We should wait for any pending migrations to complete
+			err = mig.WaitForCompletion()
```

Comment:
Are these guaranteed to eventually complete? If there's an error this function should probably timeout with an error eventually instead of just blocking forever right?

Reply:
They will as long as the lower level ReadAt() and WriteAt() calls complete - just as is the current case with migrator. I think it would be better to wrap storage with a 'timeout' module, which is then usable by anything. Will address in separate PR.